vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/SplitInstallManager.kt

```diff
@@ -0,0 +1,337 @@
+        if (triples.isNullOrEmpty()) {
+            NotificationManagerCompat.from(context).cancel(SPLIT_INSTALL_NOTIFY_ID)
+            return false
+        }
+        val intent = runCatching { installSplitPackage(context, callingPackage, triples) }.getOrNull()
+        NotificationManagerCompat.from(context).cancel(SPLIT_INSTALL_NOTIFY_ID)
+        if (intent == null) { return false }
+        sendCompleteBroad(context, callingPackage, intent)
+        return true
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private suspend fun installSplitPackage(context: Context, callingPackage: String, downloadList: ArraySet<Triple<String, String, Int>>): Intent {
+        Log.d(TAG, "installSplitPackage start ")
+        if (!context.splitSaveFile().exists()) context.splitSaveFile().mkdir()
+        val downloadSplitPackage = downloadSplitPackage(context, callingPackage, downloadList)
+        if (!downloadSplitPackage) {
+            Log.w(TAG, "installSplitPackage download failed")
+            throw RuntimeException("installSplitPackage downloadSplitPackage has error")
+        }
+        Log.d(TAG, "installSplitPackage downloaded success")
+
+        val packageInstaller = context.packageManager.packageInstaller
+        val params = PackageInstaller.SessionParams(PackageInstaller.SessionParams.MODE_INHERIT_EXISTING)
+        params.setAppPackageName(callingPackage)
+        params.setAppLabel(callingPackage + "Subcontracting")
+        params.setInstallLocation(PackageInfo.INSTALL_LOCATION_INTERNAL_ONLY)
+        try {
+            @SuppressLint("PrivateApi") val method = PackageInstaller.SessionParams::class.java.getDeclaredMethod(
+                "setDontKillApp", Boolean::class.javaPrimitiveType
+            )
+            method.invoke(params, true)
+        } catch (e: Exception) {
+            Log.w(TAG, "Error setting dontKillApp", e)
+        }
+        val sessionId: Int
+        var session: PackageInstaller.Session? = null
+        var totalDownloaded = 0L
+        try {
+            sessionId = packageInstaller.createSession(params)
+            session = packageInstaller.openSession(sessionId)
+            downloadList.forEach { item ->
+                val pkgPath = File(context.splitSaveFile().toString(), item.first)
+                session.openWrite(item.first, 0, -1).use { outputStream ->
+                    FileInputStream(pkgPath).use { inputStream -> inputStream.copyTo(outputStream) }
+                    session.fsync(outputStream)
+                }
+                totalDownloaded += pkgPath.length()
+                pkgPath.delete()
+            }
+            val deferred = CompletableDeferred<Intent>()
+            deferredMap[sessionId] = deferred
+            val intent = Intent(context, InstallResultReceiver::class.java).apply {
+                putExtra(KEY_BYTES_DOWNLOADED, totalDownloaded)
+            }
+            val pendingIntent = PendingIntent.getBroadcast(context, sessionId, intent, 0)
```

Comment:
```suggestion
            val pendingIntent = PendingIntent.getBroadcast(context, sessionId, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE)
```

Reply:
@jonathanklee With `FLAG_IMMUTABLE`, I am not receiving any response values in the `InstallResultReceiver`, so the receiver can't detect if installation was successful or not. So `FLAG_MUTABLE` seems better.